### PR TITLE
Add globbing support to dependency queries.

### PR DESCRIPTION
**Pull request by megaumi**

### CODE REVIEW

File: src/query.c [hunk 1]
```diff
@@ -980,8 +1017,18 @@ hy_query_filter(HyQuery q, int keyname, int cmp_type, const char *match)
 	HySack sack = query_sack(q);
```
Comment (jsilhan) [inline src/query.c]: What if we delete the switch, use whats in default statement and save it as a string and convert it to Reldep during `hy_query_apply`?

File: src/query.c [hunk 2]
```diff
@@ -827,23 +851,36 @@ hy_query_apply(HyQuery q)
...
+            filter_dataiterator(q, f, &m);
+        }
 	    break;
 	case HY_PKG_PROVIDES:
-	    assert(f->match_type == _HY_RELDEP);
-	    filter_provides_reldep(q, f, &m);
-	    break;
+	    if (f->match_type == _HY_RELDEP) {
+            filter_provides_reldep(q, f, &m);
+        }
```
Comment (jsilhan) [inline src/query.c]: `} else {` - this is our style guide of writing if-else statements

File: tests/python/tests/test_query.py [hunk 3]
```diff
@@ -211,6 +211,20 @@ def test_apply(self):
...
 
+    def test_provides_glob_should_work(self):
+        q1 = hawkey.Query(self.sack).filter(provides__glob="penny*")
+        self.assertLength(q1, 2)
+        q2 = hawkey.Query(self.sack).filter(provides__glob="*")
+        q3 = hawkey.Query(self.sack).filter()
+        self.assertEqual(len(q2), len(q3))
+
+    def test_provides_glob_should_not_work(self):
+        q4 = hawkey.Query(self.sack).filter(provides__glob="* >= 3")
```
Comment (jsilhan) [inline tests/python/tests/test_query.py]: I still don't see why we should not support that

File: src/query.c [hunk 4]
```diff
@@ -980,8 +1017,18 @@ hy_query_filter(HyQuery q, int keyname, int cmp_type, const char *match)
 	HySack sack = query_sack(q);
 	HyReldep reldep = reldep_from_str(sack, match);
 
-	if (reldep == NULL)
-	    return hy_query_filter_empty(q);
+	if (reldep == NULL) {
+        if (cmp_type != HY_GLOB)
+            return hy_query_filter_empty(q);
+        else {
```
Comment (jsilhan) [inline src/query.c]: it should not call `reldep_from_str` in case of HY_GLOB when we know we will not find anything

**Comment by jsilhan** [comment]
I hope we have agreed on this:
- [ ] implement `HyReldepList reldeps_from_str(HySack sack, const char *reldep_str);` which should use dataiterator for the name gotten from `parse_reldep_str` if contains glob expression.
- [ ] in `hy_query_filter` call `reldeps_from_str` in case of `HY_GLOB` and provides and pass result into `hy_query_provides_in`
- [ ] make `test_provides_glob_should_not_work` workable

Implement it this way, please. In case you have some issues don;t hesitate to ask. Thanks.

File: src/query.c (hunk 1 above)
Comment (megaumi) [inline src/query.c]: What advantage would that give us?

Comment (jsilhan) [inline src/query.c]: forget the comment. Call `reldeps_from_str` and `query_filter_reldep_in` from here. It was previous suggestion to find reldeps after it;s really needed (after `query.run()`)

File: src/query.c [hunk 5]
```diff
@@ -978,13 +980,20 @@ hy_query_filter(HyQuery q, int keyname, int cmp_type, const char *match)
...
     case HY_PKG_SUPPLEMENTS: {
 	HySack sack = query_sack(q);
-	HyReldep reldep = reldep_from_str(sack, match);
 
-	if (reldep == NULL)
-	    return hy_query_filter_empty(q);
-	int ret = hy_query_filter_reldep(q, keyname, reldep);
-	hy_reldep_free(reldep);
-	return ret;
+    if (cmp_type != HY_GLOB) {
```
Comment (jsilhan) [inline src/query.c]: (I would personally prefer non negation condition, but that's your choice.)

File: src/reldep.h [hunk 6]
```diff
@@ -39,7 +39,7 @@ void hy_reldeplist_free(HyReldepList reldeplist);
 void hy_reldeplist_add(HyReldepList reldeplist, HyReldep reldep);
 int hy_reldeplist_count(HyReldepList reldeplist);
 HyReldep hy_reldeplist_get_clone(HyReldepList reldeplist, int index);
-
+HyReldepList reldeplist_from_str(HySack sack, const char *reldep_str);
```
Comment (jsilhan) [inline src/reldep.h]: Add it to iutils.h as `reldep_from_str` please. we don't want to expose it.

**Comment by jsilhan** [comment]
This is the code I wanted, perfect job. Please move the `reldeplist_from_str` to `iutils.c`, then I will ltest it and merge.

File: src/reldep.h (hunk 6 above)
Comment (rholy) [inline src/reldep.h]: Yes, be careful about every `.h` file change. There is a big chance that any such change will upset at least Hughsie ;)

**Comment by rholy** [comment]
Jenkins, add to whitelist.

**Comment by rholy** [comment]
Jenkins, retest this please. (Actually, I need to test Jenkins, not this ;) )

**Comment by rholy** [comment]
Jenkins, retest this please.

**Comment by jsilhan** [comment]
Perfect patches with a lot of tests. Merged, thanks. (mem leak fixed in https://github.com/rpm-software-management/hawkey/commit/4968f22af0d10c55267acc615b03b80d237b19f4)

**Pull request closed without merging**